test/test_main.py

```diff
@@ -1468,19 +1468,102 @@ class TestStorageHeuristics:
+            juju_context = _JujuContext.from_dict(os.environ)
+            assert not _should_use_controller_storage(Path(fd.name), meta, juju_context)
+
+
+class TestJujuContext:
+    def test_both_str_and_int_fields_default_to_none(self):
+        juju_context = _JujuContext.from_dict({'JUJU_VERSION': '0.0.0'})
+        assert juju_context.action_name is None
+        assert juju_context.relation_id is None
+
+    def test_parsing_int_fields(self):
+        juju_context = _JujuContext.from_dict({
+            'JUJU_VERSION': '0.0.0',
+            'JUJU_RELATION_ID': 'x:42',
+        })
+        assert juju_context.relation_id == 42
+
+    def test_parsing_secret_revision_as_int(self):
+        juju_context = _JujuContext.from_dict({
+            'JUJU_VERSION': '0.0.0',
+            'JUJU_SECRET_REVISION': '42',
+        })
+        assert juju_context.secret_revision == 42
+
+    def test_parsing_juju_debug_as_bool(self):
+        juju_context = _JujuContext.from_dict({
+            'JUJU_VERSION': '0.0.0',
+            'JUJU_DEBUG': 'true',
+        })
+        assert juju_context.debug is True
+
+    def test_parsing_juju_charm_dir(self):
+        juju_context = _JujuContext.from_dict({
+            'JUJU_VERSION': '0.0.0',
+            'JUJU_CHARM_DIR': '/dir',
+        })
+        assert juju_context.charm_dir == Path('/dir')
+
+    def test_parsing_juju_charm_dir_not_set(self):
+        juju_context = _JujuContext.from_dict({'JUJU_VERSION': '0.0.0'})
+        assert juju_context.charm_dir == Path(f'{__file__}/../../..').resolve()
+
+    def test_parsing_juju_version(self):
+        juju_context = _JujuContext.from_dict({'JUJU_VERSION': '3.4.0'})
+        assert juju_context.version == JujuVersion('3.4.0')
+
+    def test_parsing_storage_id_to_name(self):
+        juju_context = _JujuContext.from_dict({
+            'JUJU_VERSION': '0.0.0',
+            'JUJU_STORAGE_ID': 'my-storage/1',
+        })
+        assert juju_context.storage_name == 'my-storage'
+
+    def test_parsing_all_str_fields(self):
+        env_var_attr_mapping = {
+            'JUJU_ACTION_NAME': 'action_name',
+            'JUJU_ACTION_UUID': 'action_uuid',
+            'JUJU_DISPATCH_PATH': 'dispatch_path',
+            'JUJU_MODEL_NAME': 'model_name',
+            'JUJU_MODEL_UUID': 'model_uuid',
+            'JUJU_NOTICE_ID': 'notice_id',
+            'JUJU_NOTICE_KEY': 'notice_key',
+            'JUJU_NOTICE_TYPE': 'notice_type',
+            'JUJU_PEBBLE_CHECK_NAME': 'pebble_check_name',
+            'JUJU_DEPARTING_UNIT': 'relation_departing_unit_name',
+            'JUJU_RELATION': 'relation_name',
+            'JUJU_REMOTE_APP': 'remote_app_name',
+            'JUJU_REMOTE_UNIT': 'remote_unit_name',
+            'JUJU_SECRET_ID': 'secret_id',
+            'JUJU_SECRET_LABEL': 'secret_label',
+            'JUJU_UNIT_NAME': 'unit_name',
+            'JUJU_WORKLOAD_NAME': 'workload_name',
```

Comment:
I'm a little confused by this - you're creating a dict that has unique values, but then in the test you only use the keys, set every value to "foo" and assert that the context attribute has the value "foo".

If you want to use "foo" for all the values, then this could just be a set rather than a dictionary. Or if you want to have unique values then it seems like you should just use `env_var_attr_mapping` and not create the `env` dict.

Reply:
I intended to maintain an environment variable to _JujuContext attribute name mapping since we renamed a few fields and they are not the same as the env vars. But I agree it is confusing, let me rewrite this test so that it's more readable. Thanks for pointing it out!